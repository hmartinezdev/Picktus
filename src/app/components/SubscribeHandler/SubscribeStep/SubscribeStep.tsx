import Button from '@components/Button';
import Input from '@components/Input';
import colors from '@constants/colors';
import { borderRadius, fontFamily } from '@constants/styles';
import { PicktusMessageLevel } from '@store/reducers/message';
import React, { ChangeEvent, Component } from 'react';

export interface ISubscribeStepPropsType {
  onValidate: (value: string, key: string) => void;
  control: (param: any) => boolean;
  errorMessage: string;
  name: string;
  title: string;
  displaySnackBar: (message: string, level: PicktusMessageLevel) => void;
}

class SubscribeStep extends Component<ISubscribeStepPropsType, IStringMap> {
  constructor(props: ISubscribeStepPropsType) {
    super(props);
    const { name } = props;

    this.state = {
      [name]: '',
    };
  }

  public onChange = (e: ChangeEvent<HTMLInputElement>): void => {
    const { value, name } = e.target;

    this.setState({ [name]: value });
  };

  public onSubmit = () => {
    const { name, onValidate, control, displaySnackBar, errorMessage } = this.props;
    const value = this.state[name];

    if (control(value)) {
      onValidate(value, name);
    } else {
      displaySnackBar(errorMessage, PicktusMessageLevel.ERROR);
    }
  };

  public render(): React.ReactElement<SubscribeStep> {
    const { name, title } = this.props;
    return (
      <div className="container">
        <h3>{title}</h3>
        <Input name={name} onChange={this.onChange} />
        <Button onClick={this.onSubmit} text="Next" />
        <style jsx>{`
          .container {
            font-family: ${fontFamily};
            color: ${colors.secondary};
            background-color: ${colors.primary};
            width: 20rem;
            padding: 1rem;
            box-sizing: border-box;
            border-radius: ${borderRadius};
          }
        `}</style>
      </div>
    );
  }
}

export default SubscribeStep;
