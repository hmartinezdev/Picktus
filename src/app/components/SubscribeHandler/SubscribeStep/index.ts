import SubscribeStep from './SubscribeStep';

export default SubscribeStep;
