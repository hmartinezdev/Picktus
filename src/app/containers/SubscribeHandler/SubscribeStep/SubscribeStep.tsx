import Button from '@components/Button';
import Input from '@components/Input';
import colors from '@constants/colors';
import { fontFamily } from '@constants/styles';
import { PicktusMessageLevel } from '@store/reducers/message';
import React, { ChangeEvent, Component } from 'react';

export interface ISubscribeStepPropsType {
  onValidate: (value: string, key: string) => void;
  control: (param: any) => boolean;
  errorMessage: string;
  name: string;
  title: string;
  displaySnackBar: (message: string, level: PicktusMessageLevel) => void;
  type?: string;
  defaultValue: string;
  current: number;
  onPreviousClick: () => void;
}

class SubscribeStep extends Component<ISubscribeStepPropsType, IStringMap> {
  public static defaultProps = {
    defaultValue: '',
    type: 'text',
  };

  constructor(props: ISubscribeStepPropsType) {
    super(props);
    const { name, defaultValue } = props;

    this.state = {
      [name]: defaultValue,
    };
  }

  public onChange = (e: ChangeEvent<HTMLInputElement>): void => {
    const { value, name } = e.target;

    this.setState({ [name]: value });
  };

  public onSubmit = () => {
    const { name, onValidate, control, displaySnackBar, errorMessage } = this.props;
    const value = this.state[name];

    if (control(value)) {
      onValidate(value, name);
    } else {
      displaySnackBar(errorMessage, PicktusMessageLevel.ERROR);
    }
  };

  public onNextClick = () => {
    this.onSubmit();
  };

  public onPreviousClick = () => {
    const { current, onPreviousClick } = this.props;
    if (current >= 1) {
      onPreviousClick();
    }
  };

  public render(): React.ReactElement<SubscribeStep> {
    const { name, title, type, defaultValue, current } = this.props;
    return (
      <div className="container">
        <Input
          name={name}
          onChange={this.onChange}
          autoFocus={true}
          placeholder={title}
          type={type}
          onEnter={this.onSubmit}
          defaultValue={defaultValue}
        />

        <div className="navigationButtons">
          <Button disabled={current < 1} onClick={this.onPreviousClick} text="previous" />
          <Button disabled={this.state[name] === ''} onClick={this.onNextClick} text="next" />
        </div>

        <style jsx>{`
          .container {
            font-family: ${fontFamily};
            color: ${colors.secondary};
          }

          :global(.navigationButtons.navigationButtons > button:nth-child(1)) {
            margin-right: 0.5rem;
          }

          :global(.navigationButtons.navigationButtons > button:nth-child(2)) {
            margin-left: 0.5rem;
          }

          .navigationButtons {
            width: 100%;
            display: flex;
            justify-content: space-around;
          }
        `}</style>
      </div>
    );
  }
}

export default SubscribeStep;
