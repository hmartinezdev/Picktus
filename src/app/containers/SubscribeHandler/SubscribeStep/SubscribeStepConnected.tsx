import { displaySnackBar } from '@store/reducers/message/message-actions';
import { connect } from 'react-redux';
import SubscribeStep from './SubscribeStep';

export default connect(
  undefined,
  { displaySnackBar }
)(SubscribeStep);
