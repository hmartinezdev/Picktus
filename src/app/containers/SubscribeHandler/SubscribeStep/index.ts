import SubscribeStepConnected from './SubscribeStepConnected';

export default SubscribeStepConnected;
