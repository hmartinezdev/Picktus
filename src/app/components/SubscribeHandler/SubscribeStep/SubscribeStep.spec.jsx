import { shallow } from 'enzyme';
import React from 'react';
import SubscribeStep from './SubscribeStep';

let wrapper;
const baseProps = {
  onValidate: () => undefined,
  control: () => true,
  errorMessage: 'error',
  name: 'step',
  title: 'stepTitle',
  displaySnackBar: () => undefined,
  type: 'text',
};
const setup = (props = {}) => shallow(<SubscribeStep {...baseProps} {...props} />);

describe('<SubscribeStep />', () => {
  test('should render properly', () => {
    wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('onSubmit', () => {
    test('it should call onValidate if control returns true', () => {
      const control = jest.fn().mockImplementation(() => true);
      const onValidate = jest.fn().mockImplementation(() => undefined);
      wrapper = setup({ control, onValidate });
      const instance = wrapper.instance();
      wrapper.setState({ [baseProps.name]: 'test' });
      wrapper.instance().onSubmit();
      expect(control).toHaveBeenCalledWith('test');
      expect(onValidate).toHaveBeenCalledWith('test', baseProps.name);
    });

    test('it should call onValidate if control returns true', () => {
      const control = jest.fn().mockImplementation(() => false);
      const displaySnackBar = jest.fn().mockImplementation(() => undefined);
      wrapper = setup({ control, displaySnackBar });
      const instance = wrapper.instance();
      wrapper.setState({ [baseProps.name]: 'test' });
      wrapper.instance().onSubmit();
      expect(control).toHaveBeenCalledWith('test');
      expect(displaySnackBar).toHaveBeenCalledWith('error', 'ERROR');
    });
  });

  describe('onChange', () => {
    test('it should call setState with the correct parameters', () => {
      wrapper = setup();
      const instance = wrapper.instance();
      const spy = jest.spyOn(instance, 'setState');
      instance.onChange({ target: { name: baseProps.name, value: 'test' } });
      expect(spy).toHaveBeenCalledWith({ [baseProps.name]: 'test' });
    });
  });
});
