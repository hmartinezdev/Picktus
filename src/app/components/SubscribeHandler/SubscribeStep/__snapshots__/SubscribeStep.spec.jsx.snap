// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscribeStep /> should render properly 1`] = `
<div
  className="container"
>
  <Input
    autoFocus={true}
    disabled={false}
    name="step"
    onChange={[Function]}
    onEnter={[Function]}
    placeholder="stepTitle"
    type="text"
  />
  <style
    jsx={true}
  >
    
          .container {
            font-family: 'Josefin Sans', sans-serif;
            color: #FAC9B8;
          }
        
  </style>
</div>
`;
